Generate documentation for request bodies

schubev opened the pull request:
The `RequestBody` decorator parses a request body and performs basic sanity checks based on the struct tags. It also generates documentation for the format of those input documents.

alexis-ld commented on ui/ui/Gruntfile.js:
```diff
@@ -101,17 +101,7 @@ module.exports = function(grunt) {
         babel: {
```
@hug33k Why keep babel in the gruntfile, do we use anywhere else than in React code ?

hug33k replied:
Yes, I use some ES2015 features such as `let` or arrow functions.

schubev committed:
req: validate structure contents based on tags

Add the 'req' package which uses structure tags to build function which
validate the contents of structures. Add the `req:"nonzero"` tag to a
structure field to indicate a zero value is invalid for that field.

schubev committed:
routes: fix inconsistency in documentation tag names

schubev committed:
routes: add decorator to get validated request body

schubev committed:
routes: decorate documentation based on input body type

schubev committed:
routes: extract context keys into their own source file

schubev committed:
routes: add the argumentKey type

schubev committed:
routes: add typesafe way to get request body

schubev committed:
routes: rename JsonRequestBody decorator to RequestBody

schubev committed:
users: use RequestBody decorator on /user/login route

schubev committed:
users: use RequestBody decorator on /user route

schubev committed:
routes: add schema to RequestBody documentation

schubev committed:
aws/s3: use RequestBody for POST /aws/billrepository

schubev committed:
aws/s3: use RequestBody for POST /aws

giubil commented on aws/routes.go:
```diff
@@ -37,6 +37,11 @@ func init() {
 			routes.RequestContentType{"application/json"},
+			routes.RequestBody{postAwsAccountRequestBody{
+				RoleArn:  "arn:aws:iam::123456789012:role/example",
+				External: "LlzrwHeiM-SGKRLPgaGbeucx_CJC@QBl,_vOEF@o",
```
Is this external random ?

schubev replied:
It is an example.

giubil replied:
Okay, that was just to make sure (the `roleArn` looked like an example, but I wasn't sure about the external)

giubil commented on routes/decoratorRequestBody.go:
```diff
@@ -0,0 +1,133 @@
+	if err != nil {
+		logger := jsonlog.DefaultLogger
+		logger.Error("Failed to build validator for type %T.", rb.Example)
+		os.Exit(1)
```
If there is a `os.Exit(1)` shouldn't the logging be `Fatal` instead of just an `Error` ?

schubev replied:
`jsonlog` doesn’t have `Fatal` yet.

giubil replied:
Okay

giubil commented on routes/decoratorRequestBody_test.go:
```diff
@@ -0,0 +1,97 @@
+	handler := H(bodyEchoHandler).With(
+		RequestBody{example},
+	)
+	bs, _ := json.Marshal(example)
```
Why don't you check for the `error` return of `json.Marshal` ?

schubev replied:
Because with `example` being a very well-defined structure which `encoding/json` will not fail with, it would only make the _test_ harder to read, with no gain.

giubil commented on routes/decoratorRequestBody_test.go:
```diff
@@ -0,0 +1,97 @@
+	handler := H(bodyEchoHandler).With(
+		RequestBody{example},
+	)
+	bs, _ := json.Marshal(example)
```
Ditto here for the `error` value

schubev replied:
Ditto.

giubil commented on util/req/req.go:
```diff
@@ -0,0 +1,3 @@
+// package req provides request body validation facilities. Using struct tags
+// you may define rules in order to validate structured user input.
+package req
```
Why is this file empty ?

schubev replied:
I didn’t know in which file to put the package documentation, so I put it here, in a file eponymous to the package, which will not contain code.

giubil commented on util/req/reqSchema.go:
```diff
@@ -0,0 +1,82 @@
+	schemaIndent = "\t"
+)
+
+func GetSchema(dst io.Writer, typ reflect.Type) error {
```
The variable name `dst` does not make sense just like this. You need to add a comment to make it more explicit

schubev replied:
`dst` is quite standard an abbreviation, and with an `io.Writer` seems fairly obvious, but I did forget to document the function.

schubev committed:
util/req: document GetSchema
